From ff4a0a439121ec65056664689b3d19d2fe7f8220 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Witold=20Bo=C5=82t?= <witold.bolt@jit.team>
Date: Sun, 27 Sep 2026 01:04:11 +0200
Subject: [PATCH 1/7] util, meson: recognise Phoenix-RTOS as a POSIX KMS/DRM
 system

Phoenix-RTOS (host_machine.system() == 'phoenix', compiler defines
__phoenix__) is a microkernel POSIX system whose DRM devices are userspace
servers reached through libdrm (like managarm, already on the list).

- meson.build: add 'phoenix' to system_has_kms_drm, so GBM/EGL-drm and the
  DRM winsyses can be enabled.
- util/detect_os.h: DETECT_OS_PHOENIX, and DETECT_OS_POSIX for it
  (without it os_time.c/os_misc.c hit "#error Unsupported OS").
- util/os_misc.c: take the plain <unistd.h> branch.
- util/u_thread.[ch]: libphoenix has no pthread_barrier_* and no
  pthread_getcpuclockid; select Mesa's own mutex+condvar barrier (as Apple
  and Haiku do) and the no-per-thread-clock path (as managarm does).
---
 meson.build          | 2 +-
 src/util/detect_os.h | 8 ++++++++
 src/util/os_misc.c   | 2 +-
 src/util/u_thread.c  | 4 ++--
 src/util/u_thread.h  | 2 +-
 5 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/meson.build b/meson.build
index d6e3a8bea26..80253d1b782 100644
--- a/meson.build
+++ b/meson.build
@@ -159,7 +159,7 @@ pre_args += '-DHAVE_OPENGL_ES_2=@0@'.format(with_gles2.to_int())
 
 with_any_opengl = with_opengl or with_gles1 or with_gles2
 
-system_has_kms_drm = ['openbsd', 'netbsd', 'freebsd', 'gnu/kfreebsd', 'dragonfly', 'linux', 'sunos', 'android', 'managarm'].contains(host_machine.system())
+system_has_kms_drm = ['openbsd', 'netbsd', 'freebsd', 'gnu/kfreebsd', 'dragonfly', 'linux', 'sunos', 'android', 'managarm', 'phoenix'].contains(host_machine.system())
 
 gallium_drivers = get_option('gallium-drivers')
 if gallium_drivers.contains('auto')
diff --git a/src/util/detect_os.h b/src/util/detect_os.h
index 74037240c74..7a870aa1674 100644
--- a/src/util/detect_os.h
+++ b/src/util/detect_os.h
@@ -28,6 +28,11 @@
 #define DETECT_OS_ANDROID 1
 #endif
 
+#if defined(__phoenix__)
+#define DETECT_OS_PHOENIX 1
+#define DETECT_OS_POSIX 1
+#endif
+
 #if defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #define DETECT_OS_FREEBSD 1
 #define DETECT_OS_BSD 1
@@ -122,6 +127,9 @@
 #ifndef DETECT_OS_HURD
 #define DETECT_OS_HURD 0
 #endif
+#ifndef DETECT_OS_PHOENIX
+#define DETECT_OS_PHOENIX 0
+#endif
 #ifndef DETECT_OS_LINUX
 #define DETECT_OS_LINUX 0
 #endif
diff --git a/src/util/os_misc.c b/src/util/os_misc.c
index 16dd51bfb08..96d49dfaf38 100644
--- a/src/util/os_misc.c
+++ b/src/util/os_misc.c
@@ -60,7 +60,7 @@
 #  include <unistd.h>
 #  include <android/log.h>
 #  include <sys/system_properties.h>
-#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM
+#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_MANAGARM || DETECT_OS_PHOENIX
 #  include <unistd.h>
 #elif DETECT_OS_OPENBSD || DETECT_OS_FREEBSD
 #  include <sys/resource.h>
diff --git a/src/util/u_thread.c b/src/util/u_thread.c
index d1cf0858c61..73e803f0533 100644
--- a/src/util/u_thread.c
+++ b/src/util/u_thread.c
@@ -176,7 +176,7 @@ util_set_thread_affinity(thrd_t thread,
 int64_t
 util_thread_get_time_nano(thrd_t thread)
 {
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__managarm__) && !defined(__phoenix__)
    struct timespec ts;
    clockid_t cid;
 
@@ -196,7 +196,7 @@ util_thread_get_time_nano(thrd_t thread)
 #endif
 }
 
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__phoenix__)
 
 void util_barrier_init(util_barrier *barrier, unsigned count)
 {
diff --git a/src/util/u_thread.h b/src/util/u_thread.h
index 57ebdc51ba1..1230b03014a 100644
--- a/src/util/u_thread.h
+++ b/src/util/u_thread.h
@@ -116,7 +116,7 @@ static inline bool u_thread_is_self(thrd_t thread)
  * util_barrier
  */
 
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__phoenix__)
 
 typedef pthread_barrier_t util_barrier;
 
